### PR TITLE
remove NSGs when upgrading masters

### LINKED ISSUE #2175: K8s upgrade leads to temporary unreachable services
**Is this a request for help?**:
---
maybe yes, as I can't believe, that upgrading K8s using acs-engine leads to temporary unavailability (connectivity loss) of deployed applications

**Is this an ISSUE or FEATURE REQUEST?** (choose one):
---
in worst case an issue

**What version of acs-engine?**:
---
v0.12.0, but also seen in v0.10.0

**Orchestrator and version (e.g. Kubernetes, DC/OS, Swarm)**
server 1.8.4 client 1.9.2

**What happened**:

On **AzureGermanCloud** I've set up a basic cluster with 3 masters and 3 agents. I run 3 instances of nginx each on every agent and having a load-balanced service to have nginx reachable from outside via public ip. I'm continuously `curl`ing to check, if nginx is reachable and it is.
Now I start to upgrade k8s from 1.8.4 to 1.9.1
When upgrade process comes to first upgrade of k8s master:

    INFO[0252] Master offset: 0
    INFO[0252] Master pool set count to: 1 temporarily during upgrade...
    INFO[0252] Starting ARM Deployment (master-18-01-31T12.25.12-930474091). This will take some time...
    [...here...]
    INFO[1171] Finished ARM Deployment (master-18-01-31T12.25.12-930474091). 

...nginx is not reachable anymore for at least 10 minutes ... curl reports connection timed out.
About 2 minutes before ARM Deployment will be finished, nginx is reachable again and keeps reachable until next k8s master gets upgraded. Then, again, nginx is not reachable, temporarily... same for 3rd master upgrade. 

If I would have to guess, the ARM Deployment "touches" (maybe re-creates) Azure's Public IP resource, which leads to the outage.

**What you expected to happen**:
I would expect that running applications, which have replicas on multiple k8s agent nodes and a load balanced service, are reachable during upgrade of k8s without interruption (as on AWS).

**How to reproduce it** (as minimally and precisely as possible):
Bootstrap at least 1 master, 2 agents, deploy nginx (or any application) with at least 2 replicas and pod affinity to be distributed to at least 2 k8s agents, setup load-balanced service and monitor app's availability via public ip connection.

**Anything else we need to know**:

    "kubernetesConfig": {
        "networkPolicy": "none"
    },

full config: https://gist.github.com/ITler/840c96544416b1ecee3add2e171f1e2b

### PR DESCRIPTION
**Pull request by yastij**
**What this PR does / why we need it**: see #2175

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #2175

**Special notes for your reviewer**:

**Release note**:

```release-note
remove NSGs when upgrading masters to be able to reach kubernetes services.
```

### CODE REVIEW

**Comment by jackfrancis** [comment]
@yastij I'll run upgrade tests against this branch once you fix lint errors (see Details link in test circleci job above)

Thanks for this!

**Comment by CecileRobertMichon** [comment]
Will this work for both VMSS and Availability Sets? cc: @sozercan

**Comment by sozercan** [comment]
@CecileRobertMichon VMSS is for agent pools only, doesn't touch the masters so should be okay.

**Comment by yastij** [comment]
@CecileRobertMichon - this is also why I extracted the NSG removal logic.

File: pkg/acsengine/transform/transform.go [hunk 1]
```diff
@@ -155,6 +155,11 @@ func (t *Transformer) NormalizeForK8sVMASScalingUp(logger *logrus.Entry, templat
 	if err := t.NormalizeMasterResourcesForScaling(logger, templateMap); err != nil {
 		return err
 	}
+	return RemoveNSGs(t, logger, templateMap)
+}
+
+// RemoveNSGs removes the NSG from the templateMap
+func RemoveNSGs(t *Transformer, logger *logrus.Entry, templateMap map[string]interface{}) error {
```
Comment (jackfrancis) [inline pkg/acsengine/transform/transform.go]: This new function doesn't just remove NSG resources, it also removes route tables and vnets. Is that the intention?

**Comment by CecileRobertMichon** [comment]
Hi @yastij, any update on this PR?

File: pkg/acsengine/transform/transform.go (hunk 1 above)
Comment (yastij) [reply pkg/acsengine/transform/transform.go]: @jackfrancis - Indeed this is just a WIP. Still need to extract vnet and route tables and ensure that NSG removal is done only once.

**Comment by yastij** [comment]
@CecileRobertMichon - Sorry I hadn't much time to look into this one. I'll be able to update it by the end of the week.

**Comment by CecileRobertMichon** [comment]
@yastij is this PR still active? I will close it for now, feel free to reopen it whenever you get the chance to move it forward

**Pull request closed without merging**